Symptoms-Checker [Desktop][Mobile]: local emergency numbers outside of US PR, Feature/1355

vsan626 opened the pull request:
### What does this pull request aim to achieve? 💯
Closes #1355 adds banner component with emergency numbers to staging
**Please confirm local emergency number works outside USA ip**
_Please be concise and link any related issue(s):_

<!--### 🚨Before submitting this pull request🚨:-->

<!--_Please do **NOT** submit this PR if you have not done the following:_-->

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix. The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".

vsan626 committed:
created static local emergency number, need to implement logic to take users location and display their local emergency number

vsan626 committed:
created static local emergency number, need to implement logic to take users location and display their local emergency number

vsan626 committed:
file containing country emergency data

vsan626 committed:
file containing country emergency data

vsan626 committed:
LocalEmergencyBanner renders emergency number based on users location working from what i see so far. would need to test for locations outside of USA. Tried to use VPN to test on local machine, but did not work

vsan626 committed:
LocalEmergencyBanner renders emergency number based on users location working from what i see so far. would need to test for locations outside of USA. Tried to use VPN to test on local machine, but did not work

vsan626 committed:
removed console.logs

vsan626 committed:
removed console.logs

vsan626 committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1355

vsan626 committed:
updated styling for emergency number and text

vsan626 committed:
updated styling for emergency number and text

joshmorel commented on client/src/components/LocalEmergencyBanner/localEmergencyData.js:
```diff
@@ -0,0 +1,1456 @@
+    "Fire": 118
+  },
+  {
+    "Country": "People's Republic of China",
```
```suggestion
    "CN": {...
```

I think we should be using standard country codes instead - see for https://extreme-ip-lookup.com/json/1.185.255.255 - the country name is "China" by "CN" is definitely standard so should be better.

We can also make it an object instead of an array so each key is a unique country code.

vsan626 replied:
Hi @joshmorel awesome! Once it's all looked over I'll go ahead and make the changes and do another pull req

vsan626 replied:
Hi @joshmorel I've updated the code and have pushed it up. Hope all is good now!

joshmorel replied:
This seems like it's working well locally with once exception (I will comment elsewhere)... too bad the review branch isn't working.

joshmorel reviewed (changes requested):
Hey @vsan626 this is definitely good work I'll check the rest soon but I think we should structure our data different and look-up by country code (see suggestion).

vsan626 committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1355

vsan626 committed:
updated emergency script

vsan626 committed:
updated emergency script

vsan626 committed:
updated code to track country's code instead of country name

vsan626 committed:
updated code to track country's code instead of country name

vsan626 committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1355

vsan626 committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1355

joshmorel commented:
Also resolves #342

joshmorel commented on client/src/components/LocalEmergencyBanner/LocalEmergencyBanner.js:
```diff
@@ -0,0 +1,96 @@
+    } else {
+        axios.get('https://extreme-ip-lookup.com/json/')
+          .then(ipInfo => {
+            let countryCode = localEmergencyData[ipInfo.data.countryCode];
```
```suggestion
            let countryCode = localEmergencyData[ipInfo.data.countryCode];
```

If for whatever reason we don't have a match (we should be just in case) , maybe we should throw an error so it displays 'location unavailable'?

joshmorel reviewed (changes requested):
Sorry for the late re-review, I was caught up in my own tricky feature.

One thing I noticed it's calling the service multiple times. We definitely want to avoid this to avoid hitting the quota.

![image](https://user-images.githubusercontent.com/10546720/90986411-acf36c80-e550-11ea-95c8-58f229bf9241.png)

I think `useEffect` with empty dependency array so it only calls this function similar to `componentWillMount`.

For similar see - https://github.com/FightPandemics/FightPandemics/blob/1f83681a862339b31b1d5d241929bc920bbf6722/client/src/pages/ProfileCompleted.js#L20

useEffect is kind of tricky to reason about if you're unfamiliar  - send me a message if you have any questions.

Later we can use `localStorage` to keep track of number on the user's browser so it's not always re-fetching.

joshmorel commented:
@vsan626 I have a few other code style suggestions.  It might be easier for me to just make a PR into your branch then comment on this PR. Hopefully tonight

joshmorel committed:
fix formatting + remove render fnc

better to return null if displayBanner false

joshmorel committed:
fix formatting + remove render fnc

better to return null if displayBanner false

joshmorel committed:
change AUS 0 emergency number to 000

joshmorel committed:
change AUS 0 emergency number to 000

joshmorel committed:
get emergency number once via useEffect

joshmorel committed:
get emergency number once via useEffect

Danny-MarketSpark committed:
Merge pull request #1414 from FightPandemics/feature/1355-j

Feature/1355 j 
@joshmorel The update looks good. I cross-checked the emergency numbers based off https://en.wikipedia.org/wiki/List_of_emergency_telephone_numbers. List was correct. I see you updated the other countries that only had 0 to 000 as well.

vsan626 committed:
Merge branch 'feature/1355' of https://github.com/FightPandemics/FightPandemics into feature/1355

joshmorel reviewed (approved):
Code wise I think we are pretty good now. Some ideas for iteration @vsan626 @tejasm2 

* To reduce external API usage (after 10k requests we are blocked or need to pay):
  * Easier & FE - We can store in localStorage their emergency number & last IP, if they load the IP again then just fetch from localStorage instead of querying the API
  * Hard & BE - Way later, maybe if usage gets really high, we can add our own caching on backend (maybe)
* When I close a banner like that, sometimes I think like cookie banner it shouldn't come back on page refresh, do you think it would be the same here (because it's not relevant to the user)? We could store a time-limited localStorage value if the close it to not show it again...

If we haven't already double checked the entire emergency number list (I haven't) I'd suggest it's worth it before deploying it.

idjevm reviewed (approved):
LGTM as well, I'd say we can merge it after there is a response on @joshmorel commnets !

vsan626 committed:
updated country codes and emergency numbers based on google search and wikipedia data

vsan626 committed:
updated country codes and emergency numbers based on google search and wikipedia data

vsan626 committed:
updated local emergency numbers with string 000 back to 0 due to if statement with typeof string. triple zeros were strings and that could have caused some errors

vsan626 committed:
updated local emergency numbers with string 000 back to 0 due to if statement with typeof string. triple zeros were strings and that could have caused some errors

vsan626 committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1355

vsan626 commented:
> Code wise I think we are pretty good now. Some ideas for iteration @vsan626 @tejasm2
> 
> * To reduce external API usage (after 10k requests we are blocked or need to pay):
>   
>   * Easier & FE - We can store in localStorage their emergency number & last IP, if they load the IP again then just fetch from localStorage instead of querying the API
>   * Hard & BE - Way later, maybe if usage gets really high, we can add our own caching on backend (maybe)
> * When I close a banner like that, sometimes I think like cookie banner it shouldn't come back on page refresh, do you think it would be the same here (because it's not relevant to the user)? We could store a time-limited localStorage value if the close it to not show it again...
> 
> If we haven't already double checked the entire emergency number list (I haven't) I'd suggest it's worth it before deploying it.

Just did a manual check and update on the country's emergency numbers by comparing them with Wikipedia & Google. Updated a few of the country codes as well. I updated country's with emergency number "000" back to 0 because of the conditional if (typeof countryCode.Ambulance === "string") *line 56*. I believe that would create some issues if we kept the triple 0's since they were typeof string. Hope the updated code is good @joshmorel. I think the suggestions from @joshmorel to reduce the external API usage and to keep the banner from coming back after being closed is good. I haven't had a chance to do either suggestions before, but can look into it. Although I'm not sure how long that will take for me to implement.

rabbi08 commented:
@vsan626 
The build URL does not reflect the changes, I append the URL as below, I can see the changes on the desktop as shown in below screenshot, but this edited URL shows blank page on the mobile.

https://feature1355.fightpandemics.xyz/symptoms-check

<img width="1088" alt="image" src="https://user-images.githubusercontent.com/67168400/91257714-a0316d00-e738-11ea-8516-18d016cd80bf.png">

vsan626 commented:
Hi @rabbi08 I'm not quite understanding. I did a quick check on mobile view based on the URL: https://feature1355.fightpandemics.xyz/symptoms-check 
Would we be able to link up and chat about it tomorrow? 
<img width="876" alt="Screen Shot 2020-08-25 at 10 14 28 PM" src="https://user-images.githubusercontent.com/11686709/91258236-c3502280-e720-11ea-866e-172bb5e93b21.png">

rabbi08 commented:
> Hi @rabbi08 I'm not quite understanding. I did a quick check on mobile view based on the URL: https://feature1355.fightpandemics.xyz/symptoms-check
> Would we be able to link up and chat about it tomorrow?
> <img alt="Screen Shot 2020-08-25 at 10 14 28 PM" width="876" src="https://user-images.githubusercontent.com/11686709/91258236-c3502280-e720-11ea-866e-172bb5e93b21.png"

@vsan626 
Surely we can chat or link up tomorrow. just adding my observation that when I'm trying to access the feature build URL or even appending the /symptom-check in the URL, Blank page is displayed. I am using iPhone 11 Pro max, P.S. the screenshot below:

![image](https://user-images.githubusercontent.com/67168400/91259692-fbfdf500-e73c-11ea-93e5-26f3c59601ea.png)

joshmorel commented:
Hey @vsan626 

> Just did a manual check and update on the country's emergency numbers by comparing them with Wikipedia & Google. Updated a few of the country codes as well. 

Thanks for that!

> I updated country's with emergency number "000" back to 0 because of the conditional if (typeof countryCode.Ambulance === "string") line 56. I believe that would create some issues if we kept the triple 0's since they were typeof string. 

The "000" shouldn't case any issues with the conditional logic. You can experiment just by changing it for your own country., e.g. for me:

```
  CA: {
    Country: "Canada",
    Police: "911",
    Ambulance: "911",
    Fire: "911",
  },
```

Works totally fine for me.

I would prefer we keep `"000"` in the mapping file in case it is used by someone else without looking at that conditional logic. Actually on further thought, they are all be treated as strings, so we could update the mapping file & then just have:

```javascript
    setLocalEmergencyNumber(countryCode.Ambulance.replace(/[/]/g, " or "));
```
Instead of:

```javascript
      if (countryCode.Ambulance === 0) {
        setLocalEmergencyNumber("000");
      } else if (typeof countryCode.Ambulance === "string") {
        setLocalEmergencyNumber(countryCode.Ambulance.replace(/[/]/g, " or "));
      } else {
        setLocalEmergencyNumber(countryCode.Ambulance);
      }
```

joshmorel commented on client/src/components/LocalEmergencyBanner/emergency-numbers-by-code.js:
```diff
@@ -0,0 +1,1406 @@
+  },
+  AU: {
+    Country: "Australia",
+    Police: 0,
```
```suggestion
    Police: "000",
```

As per the other comment I think we should have the actual full emergency "number" for all these. 

The other suggestion of changing all to strings - if you can do it relatively quickly great! But it's optional.

vsan626 replied:
@joshmorel Thanks for clarifying. I've reverted those changes back to the previous set up

vsan626 committed:
reverted emergency numbers with single 0 to 000 & reverted if logic in localEmergencyBanner

vsan626 committed:
reverted emergency numbers with single 0 to 000 & reverted if logic in localEmergencyBanner

rabbi08 commented:
@joshmorel @vsan626 
I am looking to test this PR using VPN, but the URL is not working due to 503 error.
 I re-deployed the build again by using 'Re-run all commands' still the 503 error exist.  Thanks.

joshmorel commented:
@rabbi08 I can take a look tomorrow late as dont have a comp til then. Or if we rebase this one on staging and force push it might work if you want to try @vsan626

vsan626 committed:
Merge branch 'feature/1355' of https://github.com/FightPandemics/FightPandemics into feature/1355

vsan626 commented:
@josh @rabbi08 I just did a git rebase staging then pulled any updates and pushed again. Give it a try

rabbi08 commented:
@vsan626 @joshmorel @idjevm 
I have tested this feature using WonderProxy on feature build, through VPN/Proxy server and it looks it is working fine. I have initially done test for below 3 countries and it is working fine. 

But I have Question: 
- The national emergency numbers are displayed on what defined criteria like is it, Fire Dept, National Emergency, Ambulance?

Because currently they are displayed as below for following countries:

India: displayed National Emergency number
https://share.getcloudapp.com/KourP7LE

Zurich, Switzerland: Fire Department
https://share.getcloudapp.com/yAubO0o6

Brazil: Ambulance 
https://share.getcloudapp.com/o0u827k4

galeza commented:
Hi @rabbi08 ,

I tested for France where number is displayed for Ambulance (SAMU) - 15 (this test no VPN needed, I'm based in France)
I wanted to test with WonderProxy for Warsaw. I setup System Preferences but I wasn't able to run it. I saw pop-up authentication for Warsaw server but when I entered user/password nothing happened. I must check it once more.

vsan626 commented:
> @vsan626 @joshmorel @idjevm
> I have tested this feature using WonderProxy on feature build, through VPN/Proxy server and it looks it is working fine. I have initially done test for below 3 countries and it is working fine.
> 
> But I have Question:
> 
> * The national emergency numbers are displayed on what defined criteria like is it, Fire Dept, National Emergency, Ambulance?
> 
> Because currently they are displayed as below for following countries:
> 
> India: displayed National Emergency number
> https://share.getcloudapp.com/KourP7LE
> 
> Zurich, Switzerland: Fire Department
> https://share.getcloudapp.com/yAubO0o6
> 
> Brazil: Ambulance
> https://share.getcloudapp.com/o0u827k4

@rabbi08 The emergency banner will display the ambulance numbers based on the emergency-numbers-by-code.js file once the country code has been determined from the ip lookup api

rabbi08 commented:
> Hi @rabbi08 ,
> 
> I tested for France where number is displayed for Ambulance (SAMU) - 15 (this test no VPN needed, I'm based in France)
> I wanted to test with WonderProxy for Warsaw. I setup System Preferences but I wasn't able to run it. I saw pop-up authentication for Warsaw server but when I entered user/password nothing happened. I must check it once more.

Prefer to load the webpage in incognito window as it works on cache, after VPN prefer to reopen the webpage again with no cache to see the difference after VPN connected. I will also test for France and will share the results.

rabbi08 commented:
Temporary closing this PR, but it will stay in 'Symptoms-Checker' as currently the priority for this feature is very low, hence trying to save some AWS credit by deleting the branches of low priority features.

Please feel free to re-open or create new when needed. Thanks.
